### PR TITLE
80X EGM Regression 

### PR DESCRIPTION
**Pull request by rafaellopesdesa**
This is a standalone producer so that users can apply the latest GED electron/photon regression without changing the RECO sequence in 8_0_X. The regressions are taken from the conditions database as suggested by @slava77 . We also added the option of reading from a local sqlite file. General instructions for the users are described here:

https://twiki.cern.ch/twiki/bin/view/CMS/EGMRegression

We still need a bit of work to release this product (for instance, the EGMSmearer has to be rederived for this specific regression), but the code is ready so I am doing the PR right now so that we have time to review it while these last steps are taken.

A similar PR will exist for 81X and 90X but it will modify the RECO sequence as well. However, before I do these PR, I would like that my TAGS are included in the GT. I am in contact with AlCa to add them:

https://hypernews.cern.ch/HyperNews/CMS/get/calibrations/2774.html

This PR is not supposed to have any effect in any validation step. 

Attention: @fcouderc , @ikrav , @gzevi

### CODE REVIEW

File: EgammaAnalysis/ElectronTools/plugins/EGExtraInfoModifierFromDB.cc [hunk 1]
```diff
@@ -0,0 +1,752 @@
...
+#include <vdt/vdtMath.h>
+#include <TFile.h>
+
+namespace {
+  const edm::InputTag empty_tag;
+}
+
+#include <unordered_map>
+
+class EGExtraInfoModifierFromDB : public ModifyObjectValueBase {
```
Comment (rafaellopesdesa) [inline EgammaAnalysis/ElectronTools/plugins/EGExtraInfoModifierFromDB.cc]: @slava77  Updated.

**Comment by slava77** [comment]
The old file is still in place.

I noted already in email about copy-paste from RecoEgamma.
Something to be noted by the analysis L2 in review.

**Comment by rafaellopesdesa** [comment]
That is true: the structure is *almost* the same as the plugin in RecoEgamma, but not exactly the same. The methodology of the GED regression has changed and since 80X reconstruction cannot be altered anymore, the simplest solution was to have a producer that can be run by users at analysis level. That's how we've been using EgammaAnalysis in EGM group and this PR is consistent with this use.

Again, even though the code is similar, it is not interchangeable. It is not simply a training update, but a change that would be incompatible with the current regression in the reconstruction sequence.

PS: Sorry, I forgot the git rm before. Now it is done.

**Comment by monttj** [comment]
@rafaellopesdesa 

I am wondering if this update at the analysis level can not be done with configuration file.

Having two versions in Reco and Analysis would introduce confusion to end users in my opinion.
For next open RECO release, do you plan to update the RecoEgamma accordingly? In this case, you have to maintain two versions if I understand correctly. 

Regards,
Taejeong

Regards,
Taejeong

**Comment by rafaellopesdesa** [comment]
In principle I can. But since the method changed, I would have to put a lot
of "if" statements in the code, so it would be a duplication anyway, but
only inside a file.

EgammaAnalysis has always been used to store analysis-level (user-level)
code. In 81X and 90X I will update RecoEgamma only.

**Comment by monttj** [comment]
@cmsbuild please test

**Comment by monttj** [comment]
@rafaellopesdesa 

Thank you for your answer. 

But I am sorry for my ignorance. 
I am wondering if this regression energy scale is used and saved in official miniAOD workflow?
If it is only for analysis level, this tools in RecoEgamma area can be moved to EgammaAnalysis area.

Taejeong

**Comment by slava77** [comment]
@cmsbuild please test

If I understand correctly from 90X, this PR will need more updates

**Comment by rafaellopesdesa** [comment]
That is correct. We are working on it. It will be update shortly.

**Comment by ikrav** [comment]
Slava, I cherry-picked all commits from the sister 90X PR that I had
to introduce to address your and Carl’s comments, if this is what you
mean by “updates”. If there is anything else that jumps at you that needs
to be changed, let me know.
Ilya

**Comment by slava77** [comment]
On 1/17/17 8:50 AM, ikrav wrote:
> 
> Slava, I cherry-picked all commits from the sister 90X PR that I had
> to introduce to address your and Carl’s comments, if this is what you
> mean by “updates”. If there is anything else that jumps at you that needs
> to be changed, let me know.
> Ilya

Hi Ilya,

I just wanted to check that you are commenting in the right PR.
This is for "GED electron/photon regression without changing the RECO"
If I recall correctly, your work is on IDs, separate from this PR.




>

**Comment by ikrav** [comment]
Sorry for the noise, everyone, I was commenting on the wrong PR indeed.
Please ignore.
Ilya

**Comment by rafaellopesdesa** [comment]
We had a series of updates in this regressions and code change was needed. I will close this one and do it again.

**Pull request closed without merging**